### ISSUE TITLE
Performance issues of GPU operators

### ISSUE THREAD

**Issue by bosskwei** [mindspore-ai/mindspore#126]
## use fused operation `multiply-add`/`fmad` or `addmm`
+ https://github.com/mindspore-ai/mindspore/blob/r1.1/mindspore/nn/layer/basic.py#L307
+ `nn.Dense` was implemented in very low performance, modern GPU archs have some kind of fused instructions or intrinsics, so we should follow the low level hardward isa and design a compatible high level api, that's why pytorch implemented an `addmm` instead of two-step forward.
+ Actually you can simply involke `cuda::blas::gemm`, and also rewrite backward to eliminate extra boardcast and memory access overheads.
```
x = matmul(x, weight)
if self.has_bias:
    x = bias_add(x, bias)
```

## `BiasAdd` was implemented in a wrong way
+ https://github.com/mindspore-ai/mindspore/blob/r1.1/mindspore/ccsrc/backend/kernel_compiler/gpu/math/bias_add_gpu_kernel.h#L61
+ use`cuda::blas::gemm` to compute `matmul` + `bias_add` instead of `cudnnOpTensor`, the `BiasAdd` is a simple `fadd(a, b)` operation, but `cudnnOpTensor` introduced extra `fadd( fmul(alpha, a), fmul(beta, b) )` instructions.
```
const float alpha = 1;
const float beta = 0;
CHECK_CUDNN_RET_WITH_EXCEPT(
    kernel_node_,
    cudnnOpTensor(cudnn_handle_, op_desc_, &alpha, x_desc_, x_addr, &alpha, b_desc_, b_addr, &beta, x_desc_, output_addr),
    "cudnnOpTensor failed"
);
```

## Backwards
+ Also you need to rewrite some of the backward methods for these operations I mentioned before

## `CrossEntropy` was implemented as single thread cuda program, and not numerical stable
+ https://github.com/mindspore-ai/mindspore/blob/r1.1/mindspore/ccsrc/backend/kernel_compiler/gpu/cuda_impl/cross_entropy_impl.cu#L27
+ `CrossEntropy` could be implemented better in map-reduce process with pretty good parallelism.
+ as a single thread program, other programs in a special cuda stream will always wait this single thread program to finish.
```
for (size_t i = 0; i < batch_size; ++i) {
    ...
}
```

## Problems in `CrossEntropyBackward`
+ https://github.com/mindspore-ai/mindspore/blob/r1.1/mindspore/ccsrc/backend/kernel_compiler/gpu/cuda_impl/cross_entropy_impl.cu#L45
+ GPUs do not have hardward div so div in a for loop costs a lot, and type of `batch_size` is `size_t`, different reg size from type `float`, so it seems cuda compiler did not do any optimization for div, just try `float inv_batch = 1.0 / float(batch_size);` and do multiplications.
+ the implemention is just a copy from textbook, it seems wrong and I will check it latter.
```
for (size_t i = 0; i < batch_size; i++) {
  for (size_t j = blockIdx.x * blockDim.x + threadIdx.x; j < class_num; j += blockDim.x * gridDim.x) {
    if (labels[i] == j) {
      grad[i * class_num + j] = (logits[i * class_num + j] - 1) / batch_size;
    } else {
      grad[i * class_num + j] = logits[i * class_num + j] / batch_size;
    }
  }
}
```

**Comment by zpcalan** [comment]
@bosskwei Thanks for your feedbacks and suggestions : ). 
I will reply about the problem 3 and 4.

Problem about CrossEntropy:
    The reason why we don't use multi-thread in CrossEntropyWithSparseKernel before is that only calculation of the logit and the loss of each batch can be dispatched to multiple threads.
    But the loss needs to be accumulated so we can do ReduceMean. So we use CrossEntropyWithSparseKernel to do both.
    And of course, we do notice that the performance is not that good when the batch size is too large. We are on our way of optimizing this kernel.

Problem about CrossEntropyGrad:
    We will follow your advice about div and run tests on this kernel.
    As for the correctness, our testers have run full tests on this kernel and the result should be correct. 
    You can always feedback if you find any bugs.

**Comment by WarmHouse** [comment]
Thank you for your suggestions
# use fused operation multiply-add/fmad or addmm
 MindSpore support multiple devices, including Ascend、Gpu、Cpu and mobile devices.
 Mindspore can be regarded as a compiler and is divided into front-end and back-end architecture. 
 The front-end uses basic operators to define the network structure, and then back-end may carry out different optimize strategies according to different device targets.

# BiasAdd was implemented in a wrong way
  There are some reasons for the current implementation
  - BiasAdd supports 2D for MatMul and  4D for Conv2D.  So we need a `BiasAdd` kernel. `cudnnOpTensor` supports  broadcasting in any dimension. It meet the needs of MatMul and Conv2d. 
  - In fact, we cannot implement the fusion operator with cublas gemm.
    The cublas gemm support `C= apha(A * B) + beta C`. So we should fill `bias` into C with a broadcast kernel, and then apply `C= A * B + C`. 
We will compare the performance of the two implementations later.
